### PR TITLE
Script 1.3.3

### PR DESCRIPTION
**Pull request by Itokoyamato**
- Merge PR #20: code cleaning and small optimizations
- Fix empty channels not getting removed
- Load playerList on script start
- Important performance improvements by saving the amount of data sent through NUI
- Small performance improvements by proxying nui messages through a JS script
- Add timeout on black screen display to avoid small popups on slower machines
- Merge PR #31: fix important bug in player data sharing

### CODE REVIEW

**Comment by PichotM** [comment]
Also, regarding this:
https://github.com/Itokoyamato/TokoVOIP_TS3/blob/master/fivem_script/tokovoip_script/src/c_main.lua#L197
and this
https://github.com/Itokoyamato/TokoVOIP_TS3/blob/master/fivem_script/tokovoip_script/src/c_main.lua#L203

using the serverId here makes the radio easy exploitable
you should remove them and use "source" server side

**Comment by PichotM** [comment]
btw I'm not sure using js might be better than lua
since it goes through the same API

File: fivem_script/tokovoip_script/src/c_main.lua [hunk 1]
```diff
@@ -18,9 +18,136 @@ local targetPed;
...
+	end
+	local localPos;
+
+	if useLocalPed then
+		localPos = GetPedBoneCoords(ped, HeadBone);
+	else
+		localPos = GetPedBoneCoords(targetPed, HeadBone);
+	end
+
+	for _, player in pairs(playerList) do
```
Comment (mkafrin) [inline fivem_script/tokovoip_script/src/c_main.lua]: You should be able to use a standard for loop here. GetActivePlayers() seems to return a contigous array of player ids and pairs is a lot slower. In a quick test on a local server it seems to work and saves some cpu time

Something like
```lua
for i=1, #playerList do
  local player = playerList[i];
  local playerServerId = GetPlayerServerId(player);
  ...
```

Comment (mkafrin) [inline fivem_script/tokovoip_script/src/c_main.lua]: You should be able to use a standard for loop here. GetActivePlayers() seems to return a contigous array of player ids and pairs is a lot slower. In a quick test on a local server it seems to work and saves some cpu time

Something like
```lua
for i=1, #playerList do
  local player = playerList[i];
  local playerServerId = GetPlayerServerId(player);
  ...
```

Comment (Itokoyamato) [reply fivem_script/tokovoip_script/src/c_main.lua]: While the perf difference is true, here we are working with tables with at most 100~ items. That's a rather small computing, and when I tested it in tokovoip, it did not change performance whatsoever (not even a single msec)
I still changed it out for better practice tho

**Comment by Autwey** [comment]
So now with this state. No more issues with the displayScreen, no Issues with clients not getting unmuted. Two issues i've noticed, radio channels dont give radioFX, and even though you leave the channel, debug still shows last connected

**Comment by AdrineX** [comment]
I've been noticing the radiofx bug too, also shows the channel even after you leave the channel.

**Comment by Itokoyamato** [comment]
> btw I'm not sure using js might be better than lua
> since it goes through the same API

Turns out it is actually slightly faster. On a full load it ends up not accounting for much (just a few ms) but it's still better than going straight to NUI from LUA.

> u should remove ";" since its not needed in LUA and will certainly broke the ressource

Like you said, it is optional in LUA. I prefer using semi-colons and parenthesis on conditions because I'm used to it on the several other languages I use or have used. It's easier for me to read code that way, and hopefully for some other too. It won't break anything, if at all, it could avoid issues since it tells LUA the current instruction stops at the semi-colon rather than waiting for a newline.

**Comment by Kullweet** [comment]
Are you pushing Frazzle PR #18 too for the next version ?
Good luck

**Comment by Itokoyamato** [comment]
> Are you pushing Frazzle PR #18 too for the next version ?
> Good luck

I don't have plans for PR #18 just yet. I've discussed it with him already. It appears there are some issues with stereo but that's no big deal.

The PR duplicates system in a way that's a bit too inefficient, and it also adds extra logic and calculations.
Considering the current system already has performance issues on large scale servers, it is just too perf heavy for me to include it as is.

However I do like what it's trying to add to Tokovoip and would like to look into a different implementation but haven't had the time to do so yet.

**Pull request merged**